dds/idl/ts_generator.cpp

```diff
@@ -113,36 +118,116 @@ bool ts_generator::gen_struct(AST_Structure*, UTL_ScopedName* name,
     "  typedef " << cxxName << "DataReader DataReaderType;\n"
     "  typedef " << module_scope(name) << "OpenDDSGenerated::" << short_name << "_KeyLessThan LessThanType;\n"
+    "  typedef MarshalTraits<" << cxxName << "> MarshalTraitsType;\n"
     "\n"
     "  inline static const char* type_name () { return \"" << cxxName << "\"; }\n"
-    "  inline static bool gen_has_key (const MessageType& x) { return ::OpenDDS::DCPS::gen_has_key(x); }\n"
+    "  inline static bool gen_has_key () { return " << (empty ? "false" : "true") << "; }\n"
     "\n"
-    "  inline static bool gen_is_bounded_size (const MessageType& x) { return ::OpenDDS::DCPS::gen_is_bounded_size(x); }\n"
     "  inline static size_t gen_max_marshaled_size(const MessageType& x, bool align) { return ::OpenDDS::DCPS::gen_max_marshaled_size(x, align); }\n"
     "  inline static void gen_find_size(const MessageType& arr, size_t& size, size_t& padding) { ::OpenDDS::DCPS::gen_find_size(arr, size, padding); }\n"
     "\n"
-    "  inline static bool gen_is_bounded_size (const OpenDDS::DCPS::KeyOnly<const MessageType>& x) { return ::OpenDDS::DCPS::gen_is_bounded_size(x); }\n"
     "  inline static size_t gen_max_marshaled_size(const OpenDDS::DCPS::KeyOnly<const MessageType>& x, bool align) { return ::OpenDDS::DCPS::gen_max_marshaled_size(x, align); }\n"
     "  inline static void gen_find_size(const OpenDDS::DCPS::KeyOnly<const MessageType>& arr, size_t& size, size_t& padding) { ::OpenDDS::DCPS::gen_find_size(arr, size, padding); }\n"
     "};\n}  }\n\n";
 
   {
     ScopedNamespaceGuard hGuard(name, be_global->header_);
+
     be_global->header_ <<
-      "  typedef OpenDDS::DCPS::TypeSupportImpl_T<" << cxxName << "> " << short_name << "TypeSupportImpl;\n";
+      "class " << be_global->export_macro() << " " << short_name << "TypeSupportImpl\n"
+      "  : public virtual OpenDDS::DCPS::LocalObject<" << short_name << "TypeSupport>\n"
+      "  , public virtual OpenDDS::DCPS::TypeSupportImpl\n"
+      "{\n"
+      "public:\n"
+      "  typedef OpenDDS::DCPS::DDSTraits<" << short_name << "> TraitsType;\n"
+      "  typedef " << short_name << "TypeSupport TypeSupportType;\n"
+      "  typedef " << short_name << "TypeSupport::_var_type _var_type;\n"
+      "  typedef " << short_name << "TypeSupport::_ptr_type _ptr_type;\n"
+      "\n"
+      "  " << short_name << "TypeSupportImpl() {}\n"
+      "  virtual ~" << short_name << "TypeSupportImpl() {}\n"
+      "\n"
+      "  virtual ::DDS::DataWriter_ptr create_datawriter();\n"
+      "  virtual ::DDS::DataReader_ptr create_datareader();\n"
+      "#ifndef OPENDDS_NO_MULTI_TOPIC\n"
+      "  virtual ::DDS::DataReader_ptr create_multitopic_datareader();\n"
+      "#endif /* !OPENDDS_NO_MULTI_TOPIC */\n"
+      "#ifndef OPENDDS_NO_CONTENT_SUBSCRIPTION_PROFILE\n"
+      "  virtual const OpenDDS::DCPS::MetaStruct& getMetaStructForType();\n"
+      "#endif /* !OPENDDS_NO_CONTENT_SUBSCRIPTION_PROFILE */\n"
+      "  virtual bool has_dcps_key();\n"
```

Comment:
has_dcps_key() doesn't appear in the TypeSupportImpl base class.  I wonder if this was left over and isn't needed any more.